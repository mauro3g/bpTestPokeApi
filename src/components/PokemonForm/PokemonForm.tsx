import React from 'react'
import useValues from '../../hooks/values/useValues'
import { TYPES } from '../../lib/constants/catalog'
import { FIELDS } from '../../lib/constants/fields'
import { createPokemonsService, editPokemonsService } from '../../lib/utils/api'
import { IPokemon } from '../../types/types'
import Button from '../Button/Button'

const style: { [key: string]: React.CSSProperties } = {
    modal: {
        position: 'fixed',
        zIndex: 1, /* Sit on top */
        paddingTop: '100px', /* Location of the box */
        left: 0,
        top: 0,
        width: '100%', /* Full width */
        height: '100%', /* Full height */
        overflow: 'auto', /* Enable scroll if needed */
        backgroundColor: 'rgba(0,0,0,0.4)' /* Black w/ opacity */
    },
    modalContent: {
        backgroundColor: '#fefefe',
        margin: 'auto',
        padding: '20px',
        border: '1px solid #888',
        width: '80%'
    },
    close: {
        color: '#aaaaaa',
        float: 'right',
        fontSize: '28px',
        fontWeight: 'bold',
        cursor: 'pointer'
    },
    formContent: {
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center'
    },
    inputContainer: {
        display: 'flex',
        flexDirection: 'column',
        margin: '15px'
    },
    inputBox: {
        display: 'flex',
    },
    inputLabel: {
        width: '5vw',
        margin: '5px 10px'
    },
    inputMargin: {
        marginBottom: '20px'
    }
} as const

interface Props {
    edit?: boolean
    pokemon?: IPokemon
    handleCloseForm: () => void,
    refreshPokemonData: () => Promise<void>
}
const defaultPokemon: IPokemon = {
    id: 0,
    name: '',
    image: '',
    type: 'fire',
    hp: 50,
    attack: 50,
    defense: 50,
    idAuthor: 1
}

const PokemonForm = (props: Props) => {
    const { edit, pokemon, handleCloseForm, refreshPokemonData } = props
    const { values: pokemonValues, updateValue: handleChange } = useValues(edit ? pokemon : defaultPokemon)

    const handleSubmit = async () => {
        const newPokemon: IPokemon = defaultPokemon
        if(edit){
            newPokemon.id = pokemon?.id as number
        }
        newPokemon.name = pokemonValues[FIELDS.name.key]
        newPokemon.image = pokemonValues[FIELDS.image.key]
        newPokemon.type = pokemonValues[FIELDS.type.key]
        newPokemon.hp = pokemonValues[FIELDS.hp.key]
        newPokemon.attack = pokemonValues[FIELDS.attack.key]
        newPokemon.defense = pokemonValues[FIELDS.defense.key]
        if (edit) {
            await editPokemonsService(newPokemon)
        } else {
            await createPokemonsService(newPokemon)
        }
        refreshPokemonData()
        handleCloseForm()
    }

    return (
        <div id="myModal" style={style.modal}>
            <div style={style.modalContent} >
                <span style={style.close} onClick={() => handleCloseForm()}>&times;</span>
                <form onSubmit={(e) => { e.preventDefault(); handleSubmit() }}>
                    <h4>{edit ? 'Editar Pokemon' : 'Crear Pokemon'}</h4>
                    <div style={style.formContent} >
                        <div style={style.inputContainer}>
                            <label style={style.label} htmlFor="fname">{FIELDS.name.name} :  </label>
                            <input required type="text" id="fname" value={pokemonValues[FIELDS.name.key]} onChange={(e) => handleChange(FIELDS.name.key, e.target.value)}></input><br />
                            <label style={style.label} htmlFor="fimage">{FIELDS.image.name} : </label>
                            <input required type="text" id="fimage" name="fimage" value={pokemonValues[FIELDS.image.key]} onChange={(e) => handleChange(FIELDS.image.key, e.target.value)}></input><br></br>
                            <label htmlFor="ftype">{FIELDS.type.name} </label>
                            <select name="ftype" id="ftype" value={pokemonValues[FIELDS.type.key]} onChange={(e) => handleChange(FIELDS.type.key, e.target.value)}>
                                {
                                    TYPES.map((type) => <option value={type}>{type}</option>)
                                }
                            </select>

                        </div>
                        <div style={style.inputContainer}>
                            <label style={style.label} htmlFor="fhp">{FIELDS.hp.name} : {pokemonValues[FIELDS.hp.key]}</label>
                            <input style={style.inputMargin} required type="range" min="1" max="100" id="fhp" name="fhp" value={pokemonValues[FIELDS.hp.key]} onChange={(e) => handleChange(FIELDS.hp.key, parseInt(e.target.value))}></input>
                            <label style={style.label} htmlFor="fhp">{FIELDS.attack.name} : {pokemonValues[FIELDS.attack.key]}</label>
                            <input style={style.inputMargin} required type="range" min="1" max="100" id="fhp" name="fhp" value={pokemonValues[FIELDS.attack.key]} onChange={(e) => handleChange(FIELDS.attack.key, parseInt(e.target.value))}></input>
                            <label style={style.label} htmlFor="fhp">{FIELDS.defense.name} : {pokemonValues[FIELDS.defense.key]}</label>
                            <input required type="range" min="1" max="100" id="fhp" name="fhp" value={pokemonValues[FIELDS.defense.key]} onChange={(e) => handleChange(FIELDS.defense.key, parseInt(e.target.value))}></input>
                        </div>
                    </div>
                    <Button type="submit" label={edit ? "Editar" : "Crear"}></Button>
                </form>
            </div>

        </div>
    )
}

export default PokemonForm