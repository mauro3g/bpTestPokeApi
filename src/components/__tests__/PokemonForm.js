import { render, screen, cleanup } from '@testing-library/react';
import PokemonForm from '../PokemonForm/PokemonForm'
import renderer from 'react-test-renderer'

const defaultPokemon = {
    id: 0,
    name: 'Pikachu',
    image: "https://assets.pokemon.com/assets/cms2/img/pokedex/full/012.png",
    type: 'fire',
    hp: 50,
    attack: 50,
    defense: 50,
    idAuthor: 1
}


test('should render a form component', () => {
    render(<PokemonForm handleCloseForm={()=>{}} refreshPokemonData={()=>{}} edit={true} pokemon={defaultPokemon}></PokemonForm>)
    const element = screen.getByTestId('form-component')
    expect(element).toBeInTheDocument()
    expect(element).toHaveTextContent('Editar Pokemon')
    expect(element).toHaveFormValues({
        fname: defaultPokemon.name,
        fimage: defaultPokemon.image,
        ftype: defaultPokemon.type,
        fhp: defaultPokemon.hp.toString(),
        fattack: defaultPokemon.attack.toString(),
        fdefense: defaultPokemon.defense.toString(),
    })
})

test('match snapshop', ()=>{
    const testBtn = renderer.create(<PokemonForm handleCloseForm={()=>{}} refreshPokemonData={()=>{}} edit={true} pokemon={defaultPokemon}></PokemonForm>).toJSON()
    expect(testBtn).toMatchSnapshot()
})


