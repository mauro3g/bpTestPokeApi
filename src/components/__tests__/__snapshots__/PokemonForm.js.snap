// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshop 1`] = `
<div
  id="myModal"
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0.4)",
      "height": "100%",
      "left": 0,
      "overflow": "auto",
      "paddingTop": "100px",
      "position": "fixed",
      "top": 0,
      "width": "100%",
      "zIndex": 1,
    }
  }
>
  <div
    style={
      Object {
        "backgroundColor": "#fefefe",
        "border": "1px solid #888",
        "margin": "auto",
        "padding": "20px",
        "width": "80%",
      }
    }
  >
    <span
      onClick={[Function]}
      style={
        Object {
          "color": "#aaaaaa",
          "cursor": "pointer",
          "float": "right",
          "fontSize": "28px",
          "fontWeight": "bold",
        }
      }
    >
      Ã—
    </span>
    <form
      data-testid="form-component"
      onSubmit={[Function]}
    >
      <h4>
        Editar Pokemon
      </h4>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "center",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "column",
              "margin": "15px",
            }
          }
        >
          <label
            htmlFor="fname"
          >
            Nombre
             :  
          </label>
          <input
            id="fname"
            name="fname"
            onChange={[Function]}
            required={true}
            type="text"
            value="Pikachu"
          />
          <br />
          <label
            htmlFor="fimage"
          >
            Url
             : 
          </label>
          <input
            id="fimage"
            name="fimage"
            onChange={[Function]}
            required={true}
            type="text"
            value="https://assets.pokemon.com/assets/cms2/img/pokedex/full/012.png"
          />
          <br />
          <label
            htmlFor="ftype"
          >
            Tipo
             
          </label>
          <select
            id="ftype"
            name="ftype"
            onChange={[Function]}
            value="fire"
          >
            <option
              value="fire"
            >
              fire
            </option>
            <option
              value="water"
            >
              water
            </option>
            <option
              value="normal"
            >
              normal
            </option>
            <option
              value="bug"
            >
              bug
            </option>
            <option
              value="poison"
            >
              poison
            </option>
          </select>
        </div>
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "column",
              "margin": "15px",
            }
          }
        >
          <label
            htmlFor="fhp"
          >
            Hp
             : 
            50
          </label>
          <input
            id="fhp"
            max="100"
            min="1"
            name="fhp"
            onChange={[Function]}
            required={true}
            style={
              Object {
                "marginBottom": "20px",
              }
            }
            type="range"
            value={50}
          />
          <label
            htmlFor="fattack"
          >
            Ataque
             : 
            50
          </label>
          <input
            id="fattack"
            max="100"
            min="1"
            name="fattack"
            onChange={[Function]}
            required={true}
            style={
              Object {
                "marginBottom": "20px",
              }
            }
            type="range"
            value={50}
          />
          <label
            htmlFor="fdefense"
          >
            Defensa
             : 
            50
          </label>
          <input
            id="fdefense"
            max="100"
            min="1"
            name="fdefense"
            onChange={[Function]}
            required={true}
            type="range"
            value={50}
          />
        </div>
      </div>
      <button
        data-testid="button-component"
        style={
          Object {
            "backgroundColor": "#008CBA",
            "border": "2px solid #008CBA",
            "borderRadius": "5%",
            "color": "white",
            "cursor": "pointer",
            "fontSize": "1rem",
            "margin": "2px 2px",
            "padding": "10px 32px",
            "textAlign": "center",
          }
        }
        type="submit"
      >
        Editar
      </button>
    </form>
  </div>
</div>
`;
